raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -90,470 +76,480 @@
     }
 
     @Test
     public void testInitializeAsCandidateAndBecomeLeader() throws Exception {
         final int otherNodeId = 1;
-        Set<Integer> voters = Utils.mkSet(localId, otherNodeId);
-        KafkaRaftClient client = buildClient(voters);
+        Set<Integer> voters = Utils.mkSet(LOCAL_ID, otherNodeId);
+        RaftClientTestContext context = RaftClientTestContext.build(voters);
 
-        assertEquals(ElectionState.withUnknownLeader(0, voters), quorumStateStore.readElectionState());
-        time.sleep(2 * electionTimeoutMs);
+        assertEquals(ElectionState.withUnknownLeader(0, voters), context.quorumStateStore.readElectionState());
+        context.time.sleep(2 * RaftClientTestContext.ELECTION_TIMEOUT_MS);
 
-        pollUntilSend(client);
-        assertEquals(ElectionState.withVotedCandidate(1, localId, voters), quorumStateStore.readElectionState());
+        context.pollUntilSend();
+        assertEquals(ElectionState.withVotedCandidate(1, LOCAL_ID, voters), context.quorumStateStore.readElectionState());
 
-        int correlationId = assertSentVoteRequest(1, 0, 0L);
-        deliverResponse(correlationId, otherNodeId, voteResponse(true, Optional.empty(), 1));
+        int correlationId = context.assertSentVoteRequest(1, 0, 0L);
+        context.deliverResponse(correlationId, otherNodeId, RaftClientTestContext.voteResponse(true, Optional.empty(), 1));
 
         // Become leader after receiving the vote
-        client.poll();
-        assertEquals(ElectionState.withElectedLeader(1, localId, voters), quorumStateStore.readElectionState());
-        long electionTimestamp = time.milliseconds();
+        context.client.poll();
+        assertEquals(ElectionState.withElectedLeader(1, LOCAL_ID, voters), context.quorumStateStore.readElectionState());
+        long electionTimestamp = context.time.milliseconds();
 
         // Leader change record appended
-        assertEquals(1L, log.endOffset().offset);
-        assertEquals(1L, log.lastFlushedOffset());
+        assertEquals(1L, context.log.endOffset().offset);
+        assertEquals(1L, context.log.lastFlushedOffset());
 
         // Send BeginQuorumEpoch to voters
-        client.poll();
-        assertSentBeginQuorumEpochRequest(1);
+        context.client.poll();
+        context.assertSentBeginQuorumEpochRequest(1);
 
-        Records records = log.read(0, Isolation.UNCOMMITTED).records;
+        Records records = context.log.read(0, Isolation.UNCOMMITTED).records;
         RecordBatch batch = records.batches().iterator().next();
         assertTrue(batch.isControlBatch());
 
         Record record = batch.iterator().next();
         assertEquals(electionTimestamp, record.timestamp());
-        verifyLeaderChangeMessage(localId, Collections.singletonList(otherNodeId),
-            record.key(), record.value());
+        RaftClientTestContext.verifyLeaderChangeMessage(LOCAL_ID, Collections.singletonList(otherNodeId), record.key(), record.value());
```

Comment:
@hachikuji is this a bug? Shouldn't the leader (LOCAL_ID) always be a voter (the second argument for this function)?

Reply:
This is what the LeaderChangeMessage says:
```
      {"name": "VotedIds", "type": "[]int32", "versions": "0+",
       "about": "The IDs of the voters who voted for the current leader"},
```

Reply:
Yeah, that's fair. It looks like the code current just includes all followers. I guess we need to carry over the voters into the `LeaderState` if we want to implement the description above. Let's open a separate sub-task for https://issues.apache.org/jira/browse/KAFKA-9876 and decide what we want to do there.

Reply:
Done. https://issues.apache.org/jira/browse/KAFKA-10634